mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorOps.td

```diff
@@ -1513,6 +1515,103 @@ def ExtractIterSpaceOp : SparseTensor_Op<"extract_iteration_space",
+     DeclareOpInterfaceMethods<LoopLikeOpInterface,
+      ["getInitsMutable", "getLoopResults", "getRegionIterArgs",
+       "getYieldedValuesMutable"]>,
+     DeclareOpInterfaceMethods<RegionBranchOpInterface,
+      ["getEntrySuccessorOperands"]>,
+     SingleBlockImplicitTerminator<"sparse_tensor::YieldOp">]> {
+
+  let summary = "Iterate over a sparse iteration space";
+  let description = [{
+      The `sparse_tensor.iterate` operation represents a loop (nest) over
+      the provided iteration space extracted from a specific sparse tensor.
+      The operation defines an SSA value for a sparse iterator that points
+      to the current stored element in the sparse tensor and SSA values
+      for coordinates of the stored element. The coordinates are always
+      converted to `index` type despite of the underlying sparse tensor
+      storage. When coordinates are not used, the SSA values can be skipped
+      by `_` symbols, which usually leads to simpler generated code after
+      sparsification. For example:
+
+      ```mlir
+      // The coordinate for level 0 is not used when iterating over a 2-D
+      // iteration space.
+      %sparse_tensor.iterate %iterator in %space at(_, %crd_1)
+        : !sparse_tensor.iter_space<#CSR, lvls = 0 to 2>
+      ```
+
+      `sparse_tensor.iterate` can also operate on loop-carried variables
+      and returns the final values after loop termination.
+      The initial values of the variables are passed as additional SSA operands
+      to the iterator SSA value and used coordinate SSA values mentioned
+      above. The operation region has an argument for the iterator, variadic
+      arguments for specified (used) coordiates and followed by one argument
+      for each loop-carried variable, representing the value of the variable
+      at the current iteration.
+      The body region must contain exactly one block that terminates with
+      `sparse_tensor.yield`.
+
+      `sparse_tensor.iterate` results hold the final values after the last
+      iteration. If the `sparse_tensor.iterate` defines any values, a yield
+      must be explicitly present.
+      The number and types of the `sparse_tensor.iterate` results must match
+      the initial values in the iter_args binding and the yield operands.
+
+
+      A nested `sparse_tensor.iterate` example that prints all the coordinates
+      stored in the sparse input:
+
+      ```mlir
+      func.func @nested_iterate(%sp : tensor<4x8xf32, #COO>) {
+        // Iterates over the first level of %sp
+        %l1 = sparse_tensor.extract_iteration_space %sp lvls = 0 : tensor<4x8xf32, #COO>
+        %r1 = sparse_tensor.iterate %it1 in %l1 at (%crd0)
+            : !sparse_tensor.iter_space<#COO, lvls = 0 to 1>  {
+          // Iterates over the second level of %sp
+          %l2 = sparse_tensor.extract_iteration_space %sp at %it1 lvls = 1
+              : tensor<4x8xf32, #COO>, !sparse_tensor.iterator<#COO, lvls = 0 to 1>
```

Comment:
I guess that this is a comment on the previous PR but I find it hard to understand what the types of the iterators and iteration spaces are. In the op that defines `%l1` above, you don't see it at all. Similarly, in the `iterate` on `%l1`, you don't see the `iterator` type at all. I understand that this is a trade-off but it currently makes it harder for me to understand what exactly iterators and iteration spaces are and what properties they have.

One thing I am not sure yet is what `lvls = 0 to 1` exactly means. Are those the levels that exist at all or the levels are are represented/pointed to *currently* (if there is even such a distinction). To explain: I think that `%it1` is of type `!sparse_tensor.iterator<#COO, lvls = 0 to 1>`, where you seem to have two levels (unless the upper bound is exclusive?), but I'd expect that each `iterate` only iterates over one level. So `0 to 1` can't refer to what the iterator is doing "here" but the "whole" iterator space?

Reply:
The iterator type can be obtained by directly replacing `iter_space` to `iterator` (e.g., `iter_space<#COO, lvls = 0 to 1>` -> `iterator<#COO, lvls=0 to 1>`), that was why I omit them to save space (but I can add it back if that is more clear).

`Level` is a concept that we should to describe the "physical" layers of sparse tensor storage, and yes, the upper bound of `0 to 1` is exclusive.

Reply:
Also I made the returned type from `extract_interation_space` explicit, and I do find that make the code more clear. Thanks for the suggestion.